lib/streamlit/session_state.py

```diff
@@ -0,0 +1,93 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import streamlit.report_thread as ReportThread
+from streamlit.server.server import Server
+from typing import Optional, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from streamlit.report_session import ReportSession
+
+
+class SessionState(object):
+    def __init__(self, **kwargs):
+        """SessionState is just a mechanism for users to get and set properties
+        based on their application.
+
+        Parameters
+        ----------
+        **kwargs : any
+            Default values for the session state.
+
+        Example
+        -------
+        >>> session_state = SessionState(user_name='', favorite_color='black')
+        >>> session_state.user_name = 'Mary'
+        ''
+        >>> session_state.favorite_color
+        'black'
+        """
+        for key, val in kwargs.items():
+            setattr(self, key, val)
+
+
+def get_current_session() -> "ReportSession":
+    # Getting the session id easily comes fromt he report context, which is
+    # a little weird, but a precedent that has been set.
+    ctx = ReportThread.get_report_ctx()
+    this_session: Optional["ReportSession"] = None
+
+    if ctx is not None:
+        this_session = Server.get_current().get_session_by_id(ctx.session_id)
+
+    if this_session is None:
+        raise RuntimeError(
+            "We were unable to retrieve your Streamlit session. "
+            + "Is your application utilizing threads? It's possible that could "
+            + "be conflicting with our system."
+        )
+
+    return this_session
+
+
+def get_session_state(**kwargs) -> SessionState:
+    """Gets a SessionState object for the current session.
+    Creates a new object if necessary.
+
+    Parameters
+    ----------
+    **kwargs : any
+        Default values you want to add to the session state, if we're creating a
+        new one.
+    Example
+    -------
+    >>> session_state = st.session_state(user_name='', favorite_color='black')
+    >>> session_state.user_name
+    ''
+    >>> session_state.user_name = 'Mary'
+    >>> session_state.favorite_color
+    'black'
+    Since you set user_name above, next time your script runs this will be the
+    result:
+    >>> session_state = st.session_state(user_name='', favorite_color='black')
+    >>> session_state.user_name
+    'Mary'
+    """
+    this_session = get_current_session()
+    this_session_state = this_session.get_session_state()
+    if this_session_state is None:
+        this_session_state = SessionState(**kwargs)
+        this_session.initialize_session_state(this_session_state)
```

Comment:
Not sure if this is something we consciously wanted to not do but I don't believe this syntax would be possible:
```
session_state = st.session_state(user_name='', favorite_color='black')
more_state = st.session_state(cookie='chocolate chip')
```

Reply:
That is correct. That being said, I can't envision a case where they need that. uses can set the `cookie` attribute on `session_state` and that will populate through.

Reply:
Alternatively if they wanted to use the API to update session state.  I could see users not thinking of creating a new session state attribute by setting the attribute on `session_state`?